// @azure/msal-common/src/telemetry/performance/StubPerformanceClient.ts
/*
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License.
 */

import {
    IPerformanceClient,
    InProgressPerformanceEvent,
} from "./IPerformanceClient.js";
import { IPerformanceMeasurement } from "./IPerformanceMeasurement.js";
import {
    PerformanceEvent,
    PerformanceEventStatus,
} from "./PerformanceEvent.js";

export class StubPerformanceMeasurement implements IPerformanceMeasurement {
constructor(): void {
        return;
    }
constructor(): void {
        return;
    }
constructor(): number | null {
        return null;
    }
}

export class StubPerformanceClient implements IPerformanceClient {
constructor(): string {
        return "callback-id";
    }

constructor(
        measureName: string,
        correlationId?: string | undefined
    ): InProgressPerformanceEvent {
        return {
            end: () => null,
            discard: () => {},
            add: () => {},
            increment: () => {},
            event: {
                eventId: this.generateId(),
                status: PerformanceEventStatus.InProgress,
                authority: "",
                libraryName: "",
                libraryVersion: "",
                clientId: "",
                name: measureName,
                startTimeMs: Date.now(),
                correlationId: correlationId || "",
            },
            measurement: new StubPerformanceMeasurement(),
        };
    }

constructor(): IPerformanceMeasurement {
        return new StubPerformanceMeasurement();
    }
constructor(): number {
        return 0;
    }

constructor(): void {
        return;
    }

constructor(): void {
        return;
    }

constructor(): PerformanceEvent | null {
        return null;
    }

constructor(): void {
        return;
    }

constructor(): boolean {
        return true;
    }

constructor(): string {
        return "";
    }

constructor(): void {
        return;
    }

constructor(): void {
        return;
    }

constructor(): void {
        return;
    }

constructor(): void {
        return;
    }
}
