// @azure/msal-common/src/broker/nativeBroker/INativeBrokerPlugin.ts
/*
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License.
 */

import { AccountInfo } from "../../account/AccountInfo.js";
import { LoggerOptions } from "../../config/ClientConfiguration.js";
import { NativeRequest } from "../../request/NativeRequest.js";
import { NativeSignOutRequest } from "../../request/NativeSignOutRequest.js";
import { AuthenticationResult } from "../../response/AuthenticationResult.js";

export interface INativeBrokerPlugin {
    isBrokerAvailable: boolean;
constructor(loggerOptions: LoggerOptions): void;
constructor(
        accountId: string,
        correlationId: string
    ): Promise<AccountInfo>;
constructor(
        clientId: string,
        correlationId: string
    ): Promise<AccountInfo[]>;
constructor(request: NativeRequest): Promise<AuthenticationResult>;
constructor(
        request: NativeRequest,
        windowHandle?: Buffer
    ): Promise<AuthenticationResult>;
constructor(request: NativeSignOutRequest): Promise<void>;
}
